Update 1.10 enterprise upgrade docs. (#10446)
hashicorp/consul: website/content/docs/upgrading/instructions/upgrade-to-1-10-x.mdx

`CONSUL_LICENSE_PATH` environment variable or the `CONSUL_LICENSE` 
environment variable. See the [licensing documentation](/docs/enterprise#licensing) 
for more information about how to configure the license.

**5.** Upgrade all client agents to the latest 1.8.x or 1.9.x release.

**6.** Pre-configure the client agents with a license.

This can either be set with the `license_path` configuration item, the 
`CONSUL_LICENSE_PATH` environment variable or the `CONSUL_LICENSE` 
environment variable. See the [licensing documentation](/docs/enterprise#licensing) 
for more information about how to configure the license.

**7.** Upgrade all snapshot agents to the latest 1.8.x or 1.9.x release.

**8.** Pre-configure the server agents with a license.

This can either be set with the `license_path` configuration item, the 
`CONSUL_LICENSE_PATH` environment variable or the `CONSUL_LICENSE` 
environment variable. See the [licensing documentation](/docs/enterprise#licensing) 
for more information about how to configure the license.

**9.** Upgrade all server agents to v1.10.0

**10.** Upgrade all client agents to v1.10.0

**11.** Upgrade all snapshot agents to v1.10.0
</Tab>
</Tabs>

> ## Post-Upgrade Configuration Changes

Reviewer: Let's remove this section, its confusing to add it and then say nothing further is required.

Author: It seemed part of the "template" for these pages. I agree it does not seem useful but this page about upgrading to 1.8.x has the exact same bit about nothing extra being required: https://www.consul.io/docs/upgrading/instructions/upgrade-to-1-8-x

Reviewer: same as above. fine by me to leave it as is if its a shared format.

---
layout: docs
page_title: Upgrading to 1.10.0
description: >-
  Specific versions of Consul may have additional information about the upgrade
  process beyond the standard flow.
---

# Upgrading to 1.10.0

<EnterpriseAlert />

## Introduction

This guide explains how to best upgrade a single Consul Enterprise datacenter to v1.10.0
from a version of Consul that is forwards compatible with v1.10. If you are on a major 

> version of Consul prior to 1.8, please follow the link for the version you are on from 
> [here](/docs/upgrading/instructions). If you are already on major versions 1.8 or 1.9

Reviewer 2: I wonder if this is clear enough that they should follow that other guide _and then come back_ and complete this one? Perhaps something like:

```suggestion
version of Consul prior to 1.8, you will need to complete and [upgrade to 1.8.13](/docs/upgrading/instructions) or higher before continuing with this guide. If you are already on major versions 1.8 or 1.9
```

this guide will go over the procedures required for upgrading to v1.10. If you have
multiple Consul datacenters then they are upgraded in the normal order with each
upgrade taking into account the instructions below.

For the open source version of Consul please follow the 
[General Upgrade Process](/docs/upgrading/instructions/general-process).

~> You can only upgrade to Consul Enterprise 1.10 from a version of Consul Enterprise 
1.8 >= 1.8.13 or 1.9 >= 1.9.7. Other versions of Consul Enterprise are not forwards

> compatible with v1.10 and will cause issues during the upgrade that could result in
> agents failing to start.

Reviewer 2: ```suggestion
compatible with v1.10 and will cause issues during the upgrade that could result in
client and snapshot agents failing to start due to [changes in the way we manage licenses](/??).
```
Not sure offhand where that link should go but maybe the FAQ on learn?

Reviewer 2: Also it would be worth adding something like this to the callout box:

> This guide will walk through upgrading first to a compatible patch release and then completing the 1.10 upgrade from there.

Without that, I even was confused that right after saying "you need 1.8.13" the next paragraph below says " you need >= 1.8.0" before I realised why the discrepancy.

Author: I think I addressed both bits of this. I tried to make it clearer that this guide can really only be used if you are using versions 1.8+ and if you are not you should follow the normal upgrade procedures to get there first.

## Requirements

- All Consul servers, clients and snapshot agents should be on a version of Consul >= 1.8.0 and < 1.10.0.

## Assumptions

> This guides makes the following assumptions:

Reviewer: I would label this section header `#Prerequisites` and then list the prerequisites below.

Author: That would make more sense to me. However this format seemed standardized across all the upgrade instructions linked on this page https://www.consul.io/docs/upgrading/instructions.

Do you still think its worthwhile deviating in just this doc? Maybe its a change we want to make across all of them?

Reviewer: ah I didn't realize this was some shared format, ok to leave this as is for now, it was a small nit.

- You are familiar with the [General Upgrade Process](/docs/upgrading/instructions/general-process).

> - You have the ability to run Consul CLI commands. 

Reviewer 2: Do we need to state that they also have an operator token?

Author: Maybe a conditional requirement if ACLs are in use?

## Considerations

The main breaking changes in Consul Enterprise 1.10 that needs special handling during upgrading are
the licensing changes outlined on the [Specific Version Details](/docs/upgrading/upgrade-specific#consul-1-10-0)

> page. There are notes on that paged about other changes that might cause issues during an upgrade as well.

Reviewer 2: ```suggestion
page. There are notes on that page about other changes that might cause issues during an upgrade as well.
```

You can find more granular details in the full [changelog](https://github.com/hashicorp/consul/blob/release/release/1.10.0/CHANGELOG.md#1100-june-22-2021).

> Looking through these changes prior to upgrading is highly recommended.

Reviewer 2: Worth linking the FAQ too?

## Procedures

**1.** Retrieve the datacenter's current license by executing the following CLI command:

```shell
consul license get -signed
```

This will be useful to have on hand when preparing the license for the 1.10 upgrades later

> on in the process.

Reviewer: Does this save the license in a file named LICENSE? Would be good to explicitly state that since the rest of this guide is very specific and operators can follow the instructions to the T.

Author: I changed the command to save a `consul.hclic` file.

**2.** Determine if ACLs are enabled by running the following CLI command.

```shell
consul info | grep "acl ="
````

If you receive output like either of the following then ACLs are enabled:

```text
Error querying agent: Unexpected response code: 403 (Permission denied)
```

```text
        acl = enabled

```

<Tabs>

> <Tab heading="ACLs Enabled">

Reviewer 2: I like the tabs approach which limits the apparent complexity. Perhaps we should make double sure this is easy to follow with a sentence above like:

> Select the correct tab below to see the steps based on whether you have ACLs enabled of disabled in your environment.

Reviewer 3: Does it make sense to have a separate tab for Kubernetes? The upgrade path for Kubernetes will read slightly differently and I can either add it here as a suggestion or branch off of this PR branch and document it there.

Reviewer 3: We wouldnt need a separate tab for ACLs enabled vs disabled for Kubernetes as the helm chart manages that complexity under the hood for the user. They only need to provide us with the license file and we ensure we mount it to the right components depending on their existing configuration.

Reviewer 4: > They only need to provide us with the license file and we ensure we mount it to the right components depending on their existing configuration.

In fact users will already have a license file in a secret so they only need to bump the Consul image and then run `helm upgrade`. Basically we should copy this text: https://www.consul.io/docs/enterprise/license/faq#kubernetes somewhere in here.

Reviewer 3: @lkysow We created a tab for Kubernetes and have described just that. We could add the faq steps there as well. The only additional thing we added were steps to ensure people who had consul servers external to kubernetes had steps that they could follow as well. https://consul-m514xwknh-hashicorp.vercel.app/docs/upgrading/instructions/upgrade-to-1-10-x

Reviewer 4: perfect

> **3.** Upgrade all server agents to the latest 1.8.x or 1.9.x release.

Reviewer 2: ```suggestion
**3.** Upgrade all server agents to the latest 1.8.x or 1.9.x release. As with standard upgrades you should follow our [General Update Process](/docs/upgrading/instructions/general-process)
``` 

maybe the same below just to be totally clear?

Author: Do you think we still need this now that I added a callout at the very beginning of the Procedures section?

**4.** Upgrade all client agents to the latest 1.8.x or 1.9.x release.

> **5.** Upgrade all snapshot agents to the latest 1.8.x or 1.9.x release.

Reviewer: maybe link to https://www.consul.io/commands/snapshot/agent for the "Snapshot agents" phrase?

**6.** Pre-configure the server agents with a license.

This can either be set with the `license_path` configuration item, the 
`CONSUL_LICENSE_PATH` environment variable or the `CONSUL_LICENSE` 
environment variable. See the [licensing documentation](/docs/enterprise#licensing) 
for more information about how to configure the license.

**7.** Upgrade all server agents to v1.10.0

**8.** Upgrade all client agents to v1.10.0

**9.** Upgrade all snapshot agents to v1.10.0

</Tab>
<Tab heading="ACLs Disabled">

> 

Reviewer 2: Is it worth an introductory sentence here pointing out that we don't recommend running in production with ACLs disabled and that an alternative path is to work on enabling ACLs prior to upgrading?

Reviewer: :+1 ,  I think its worth stating that and linking to our ACL guidel

**3.** Upgrade all server agents to the latest 1.8.x or 1.9.x release.

**4.** Pre-configure the client agents with a license.

This can either be set with the `license_path` configuration item, the 
`CONSUL_LICENSE_PATH` environment variable or the `CONSUL_LICENSE` 
environment variable. See the [licensing documentation](/docs/enterprise#licensing) 

> for more information about how to configure the license.

Reviewer 2: Should we note that if they are adding this to the agent config file, they need to be careful to complete the upgrade immediately otherwise client agents will fail to parse the config on restarts if they aren't upgraded yet.

Author: I added something noting there intended use, that we don't support online reloading and that it is important to promptly proceed with the upgrade and not leave agents in this state forever.

> **5.** Upgrade all client agents to the latest 1.8.x or 1.9.x release.

Reviewer 2: feels worth calling out the normal processes for checking breaking changes and taking snapshots etc must be followed here?

Author: Specifically here? We call out checking for breaking changes in the Considerations section already.

I can add the bit about snapshotting but where should it be. I could see us advocating for taking a snapshot at each point where versions are stable (pre-upgrade and 1.8.13/1.9.7 upgrade is complete across servers, clients and snapshot agents)

Author: Given our conversation I think we decided it is okay as is after my modifications to add in information about when to take snapshots.

> **6.** Pre-configure the client agents with a license.

Reviewer 2: This is the same as step #4 ? Should it be here or there?

Reviewer 2: Or was this meant to read "snapshot agents"?

Author: It was meant to read snapshot agents.

`CONSUL_LICENSE_PATH` environment variable or the `CONSUL_LICENSE` 
environment variable. See the [licensing documentation](/docs/enterprise#licensing) 
for more information about how to configure the license. Note that while
Consul Enterprise 1.8.13 and 1.9.7 have the ability to load the license
from these configurations, it is intended to only be used during upgrades.
Licenses loaded this way are not online reloadable. It is therefore important
to promptly proceed with the upgrade and not leave agents in the intermediate state.

**5.** Upgrade all client agents to the latest 1.8.x or 1.9.x release.

**6.** Pre-configure the snapshot agents with a license. This is the same process that
was executed for the servers in step 4.

**7.** Upgrade all [snapshot agents](/docs/commands/snapshot/agent) to the latest 1.8.x or 1.9.x release.

**8.** Pre-configure the server agents with a license. This is the same process that
was executed for the servers in step 4.

**9.** Upgrade all server agents to v1.10.0.

**10.** Upgrade all client agents to v1.10.0.

**11.** Upgrade all [snapshot agents](/docs/commands/snapshot/agent) to v1.10.0.

> </Tab>

Reviewer 3: Tentative Kubernetes Docs
```suggestion
</Tab>
<Tab heading="Kubernetes">
  
**3.** Upgrade to the latest Consul helm chart v0.32.0+.

**4.** Update the value of `global.image` in the values file to the latest 1.8.x or 1.9.x image.
Additionally, ensure that the Kuberenetes secret with the license is correctly specified in the
values `server.enterpriseLicense.secretName` and `server.enterpriseLicense.secretKey`.

**5.** Upgrade the cluster.
  
Follow the steps in the [Kubernetes Upgrade Guide](docs/k8s/upgrade#upgrade-consul-on-kubernetes)
to then upgrade the cluster. Ensure you check the steps to upgrade the [servers](docs/k8s/upgrade#upgrading-consul-servers)
and the [clients](https://www.consul.io/docs/k8s/upgrade#upgrading-consul-clients).

**6.** Update the value of `global.image` in the values file to the 1.10.0 image.

**7.** Upgrade the cluster.

Follow the steps in the [Kubernetes Upgrade Guide](docs/k8s/upgrade#upgrade-consul-on-kubernetes)
to then upgrade the cluster. Ensure you check the steps to upgrade the [servers](docs/k8s/upgrade#upgrading-consul-servers)
and the [clients](https://www.consul.io/docs/k8s/upgrade#upgrading-consul-clients).
  
</Tab>
```

Reviewer 3: https://github.com/hashicorp/consul/pull/10449

Author: What about if only clients are in k8s?

compatible with v1.10 and will cause issues during the upgrade that could result in
agents failing to start due to [changes in the way we manage licenses](/docs/enterprise/license/faq).

## Requirements

- All Consul servers, clients and snapshot agents should be on a version of Consul >= 1.8.0 and < 1.10.0. If
  they are not at the minimum version or higher, follow [normal upgrade procedures](/docs/upgrading) to get them there.

## Assumptions

This guides makes the following assumptions:

- You are familiar with the [General Upgrade Process](/docs/upgrading/instructions/general-process).
- You have the ability to run Consul CLI commands. 
- If ACLs are in use then you possess a token with at least `operator:read` permissions.

## Considerations

The main breaking changes in Consul Enterprise 1.10 that need special handling during upgrading are
the licensing changes outlined on the [Specific Version Details](/docs/upgrading/upgrade-specific#consul-1-10-0)
page. There are notes on that page about other changes that might cause issues during an upgrade as well.
You can find more granular details in the [licensing FAQ](/docs/enterprise/license/faq) as well as the full 

> [changelog](https://github.com/hashicorp/consul/blob/master/CHANGELOG.md#1100-june-22-2021).

Author: Note that this link will be broken until we release 1.10

Looking through these changes prior to upgrading is highly recommended.

## Procedures

> ~> When any of the steps below call for upgrading components, we assume you follow the normal safe upgrade

Reviewer: nit, suggested re-wording 
Any steps in the following section that mention upgrading servers/clients assume that normal safe upgrade procedures are followed. ..

Author: 👍

```

  This will save the license to a file in the local directory named `consul.hclic`. Having this 
  will be useful when preparing the license for the 1.10 upgrades later on in the process.

**2.** Take a snapshot of the cluster by running:

```shell
consul snapshot save original.snap
```

  This snapshot should not be needed but if something were to go wrong, having
  the snapshot would make it possible to restore the previous state.

**3.** Determine if ACLs are enabled by running the following CLI command.

```shell
consul info | grep "acl ="
````

  If you receive output like either of the following then ACLs are enabled:

```text
Error querying agent: Unexpected response code: 403 (Permission denied)
```

```text
        acl = enabled

```

  Select the correct tab below based on whether ACLs are enabled or disabled or if you
  are operating Consul within Kubernetes.

<Tabs>
<Tab heading="ACLs Enabled">

**4.** Upgrade all server agents to the latest 1.8.x or 1.9.x release.

**5.** Upgrade all client agents to the latest 1.8.x or 1.9.x release.

**6.** Upgrade all [snapshot agents](/docs/commands/snapshot/agent) to the latest 1.8.x or 1.9.x release.

> **7.** Take a snapshot of the upgraded cluster by running:

Reviewer 2: nit: where should they run that, on a server? Should we mention checking disk space before running it? That gets technical but maybe at least a nod towards the fact this could write a large file to disk might be useful?

Author: I added information that said they could be run from pretty much anywhere but if you don't execute from a server/client then you will need to pass other options to the save command (and linked out to it). Also added a second note about disk requirements.

  values `server.enterpriseLicense.secretName` and `server.enterpriseLicense.secretKey`.

**7.** Upgrade the cluster.

  Follow the steps in the [Kubernetes Upgrade Guide](/docs/k8s/upgrade#upgrade-consul-on-kubernetes)
  to then upgrade the cluster. Ensure you check the steps to upgrade the [servers](/docs/k8s/upgrade#upgrading-consul-servers)
  and the [clients](/docs/k8s/upgrade#upgrading-consul-clients).

**8.** If the Consul servers are running outside the Kubernetes cluster, ensure they are upgraded
  to 1.10.0 before the next steps. 
  
  This will require pre-configuring the servers with a license
  before running 1.10. This can either be set with the `license_path` configuration item, the 
  `CONSUL_LICENSE_PATH` environment variable or the `CONSUL_LICENSE` 
  environment variable. See the [licensing documentation](/docs/enterprise/license/overview) 
  for more information about how to configure the license. Note that while
  Consul Enterprise 1.8.13 and 1.9.7 have the ability to load the license
  from these configurations, it is intended to only be used during upgrades.
  Licenses loaded this way are not online reloadable. It is therefore important

>   to promptly proceed with the upgrade and not leave agents in the intermediate state.

Reviewer 2: Should this block be here in the Kube guide @thisisnotashwin? It seems odd talking about ENV vars when the use only needs to set things in helm values?

Reviewer 2: never mind this is for external servers case

- You are familiar with the [General Upgrade Process](/docs/upgrading/instructions/general-process).
- You have the ability to run Consul CLI commands. 
- If ACLs are in use then you possess a token with at least `operator:read` permissions.

## Considerations

The main breaking changes in Consul Enterprise 1.10 that needs special handling during upgrading are
the licensing changes outlined on the [Specific Version Details](/docs/upgrading/upgrade-specific#consul-1-10-0)
page. There are notes on that paged about other changes that might cause issues during an upgrade as well.
You can find more granular details in [FAQ](/docs/enterprise/license/faq) as well as the full 
[changelog](https://github.com/hashicorp/consul/blob/release/release/1.10.0/CHANGELOG.md#1100-june-22-2021).
Looking through these changes prior to upgrading is highly recommended.

## Procedures

~> When any of the steps below call for upgrading components, we assume you follow
the normal safe upgrade procedure you normally use for upgrades. The basic server

> upgrade procedure is documented [here](/docs/upgrading/instructions/general-process).

Reviewer: Nit, instead of linking to the word "here" suggest rewording the sentence to "Refer to our documentation for [basic server upgrade process](), and for [autopilot assisted upgrades]()"

Author: Done.
